tests/test_e2e.py

```diff
@@ -0,0 +1,329 @@
+        self.assertEqual(
+            result_from_wire['access_token'], result_from_cache['access_token'],
+            "We should get a cached AT")
+
+    def _test_username_password(self,
+            authority=None, client_id=None, username=None, password=None, scope=None,
+            **ignored):
+        assert authority and client_id and username and password and scope
+        self.app = msal.PublicClientApplication(client_id, authority=authority)
+        result = self.app.acquire_token_by_username_password(
+            username, password, scopes=scope)
+        self.assertLoosely(result)
+        # self.assertEqual(None, result.get("error"), str(result))
+        self.assertCacheWorksForUser(result, scope, username=username)
+
+
+THIS_FOLDER = os.path.dirname(__file__)
+CONFIG = os.path.join(THIS_FOLDER, "config.json")
+@unittest.skipUnless(os.path.exists(CONFIG), "Optional %s not found" % CONFIG)
+class FileBasedTestCase(E2eTestCase):
+    # This covers scenarios that are not currently available for test automation.
+    # So they mean to be run on maintainer's machine for semi-automated tests.
+
+    @classmethod
+    def setUpClass(cls):
+        with open(CONFIG) as f:
+            cls.config = json.load(f)
+
+    def skipUnlessWithConfig(self, fields):
+        for field in fields:
+            if field not in self.config:
+                self.skipTest('Skipping due to lack of configuration "%s"' % field)
+
+    def test_username_password(self):
+        self.skipUnlessWithConfig(["client_id", "username", "password", "scope"])
+        self._test_username_password(**self.config)
+
+    def test_auth_code(self):
+        self.skipUnlessWithConfig(["client_id", "scope"])
+        from msal.oauth2cli.authcode import obtain_auth_code
+        self.app = msal.ClientApplication(
+            self.config["client_id"],
+            client_credential=self.config.get("client_secret"),
+            authority=self.config.get("authority"))
+        port = self.config.get("listen_port", 44331)
+        redirect_uri = "http://localhost:%s" % port
+        auth_request_uri = self.app.get_authorization_request_url(
+            self.config["scope"], redirect_uri=redirect_uri)
+        ac = obtain_auth_code(port, auth_uri=auth_request_uri)
+        self.assertNotEqual(ac, None)
+
+        result = self.app.acquire_token_by_authorization_code(
+            ac, self.config["scope"], redirect_uri=redirect_uri)
+        logger.debug("%s.cache = %s",
+            self.id(), json.dumps(self.app.token_cache._cache, indent=4))
+        self.assertIn(
+            "access_token", result,
+            "{error}: {error_description}".format(
+                # Note: No interpolation here, cause error won't always present
+                error=result.get("error"),
+                error_description=result.get("error_description")))
+        self.assertCacheWorksForUser(result, self.config["scope"], username=None)
+
+    def test_client_secret(self):
```

Comment:
Can we add this one in the set of automation tests using lab api ?

Reply:
Good point! Sum up our offline conversation:

* Currently we implicitly tested the confidential client with secret scenario, by becoming a confidential client to call lab user api and lab password api
* There seems no "app api" from the lab available at this time.
* In some future effort (as a separated PR), we will eventually get there, to cover even more than confidential client with secret, such as confidential client with certificate, and Subject Name Issuer (SNI) scenario. But right now we won't let this PR to block our upcoming other feature PRs.